File: src/dmp/DmpContextualTwoStep.cpp
/**
 * @file DmpContextualTwoStep.cpp
 * @brief  Contextual Dmp class source file.
 * @author Freek Stulp
 *
 * This file is part of DmpBbo, a set of libraries and programs for the 
 * black-box optimization of dynamical movement primitives.
 * Copyright (C) 2014 Freek Stulp, ENSTA-ParisTech
 * 
 * DmpBbo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 * 
 * DmpBbo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with DmpBbo.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "dmp/Dmp.hpp"
#include "dmp/DmpContextualTwoStep.hpp"

#include "dmp/Trajectory.hpp"
#include "functionapproximators/FunctionApproximator.hpp"
#include "functionapproximators/FunctionApproximatorLWR.hpp"

#include <boost/serialization/vector.hpp>

#include <iostream>
#include <eigen3/Eigen/Core>

using namespace std;
using namespace Eigen;

namespace DmpBbo {

DmpContextualTwoStep::DmpContextualTwoStep(int n_dims_dmp, std::vector<FunctionApproximator*> function_approximators, FunctionApproximator* policy_parameter_function, DmpType dmp_type) 
:  DmpContextual(n_dims_dmp, function_approximators, dmp_type)
{
  policy_parameter_function_ = vector<vector<FunctionApproximator*> >(dim_orig());
  for (int dd=0; dd<dim_orig(); dd++)
  { 
    policy_parameter_function_[dd] = vector<FunctionApproximator*>(1);
    policy_parameter_function_[dd][0] = policy_parameter_function->clone();
  }
}

// Overloads in DMP computeFunctionApproximatorOutput
void DmpContextualTwoStep::computeFunctionApproximatorOutput(const MatrixXd& phase_state, MatrixXd& fa_output) const
{
  int n_time_steps = phase_state.rows(); 
  fa_output.resize(n_time_steps,dim_orig());
  fa_output.fill(0.0);
  
  if (task_parameters_.rows()==0)
  {
    // When the task parameters are not set, we cannot compute the output of the function approximator.
    return;
  }
  
  MatrixXd task_parameters = task_parameters_;
  if (task_parameters.rows()==1)
  { 
    task_parameters = task_parameters.row(0).replicate(n_time_steps,1).eval();
  }
  else if (task_parameters.cols()==1)
  {
    task_parameters = task_parameters.col(0).transpose().replicate(n_time_steps,1).eval();
  }

  assert(n_time_steps==task_parameters.rows());
  
  //int n_task_parameters = task_parameters.cols();
  
  VectorXd model_parameters;
  MatrixXd output(1,1);
  for (int dd=0; dd<dim_orig(); dd++)
  { 
    int n_parameters = function_approximator(dd)->getParameterVectorSelectedSize();
    model_parameters.resize(n_parameters);
    for (int pp=0; pp<n_parameters; pp++)
    {
      policy_parameter_function_[dd][pp]->predict(task_parameters,output);
      model_parameters[pp] = output(0,0);
    }
    function_approximator(dd)->setParameterVectorSelected(model_parameters);
  }

  // The parameters of the function_approximators have been set, get their outputs now.  
  for (int dd=0; dd<dim_orig(); dd++)
  {
    function_approximator(dd)->predict(phase_state,output);
    if (output.size()>0)
    {
      fa_output.col(dd) = output;
    }
  }

}

bool DmpContextualTwoStep::isTrained(void) const
{
  for (int dd=0; dd<dim_orig(); dd++)
    for (unsigned int pp=0; pp<policy_parameter_function_[dd].size(); pp++)
      if (!policy_parameter_function_[dd][pp]->isTrained())
        return false;
      
  return true;   
}

void  DmpContextualTwoStep::train(const vector<Trajectory>& trajectories, const vector<MatrixXd>& task_parameters, string save_directory, bool overwrite)
{
  // Here's the basic structure of this function
  // 1) Do some checks
  // 2) Train a separate Dmp for each demonstration, and get the resulting model parameters
  // 3) Gather all task parameter values for all demonstrations
  // 4) Train the policy parameter function for each dimension and each model parameter
  
  
  //-----------------------------------------------------
  // 1) Do some checks
  
  // Check if inputs are of the right size.
  unsigned int n_demonstrations = trajectories.size();
  assert(n_demonstrations==task_parameters.size());  
  
  
  // Then check if the trajectories have the same duration and initial/final state
  // Later on, if they are not the same, they should be learned also.
  checkTrainTrajectories(trajectories);

  // Set tau, initial_state and attractor_state from the trajectories 
  set_tau(trajectories[0].duration());
  set_initial_state(trajectories[0].initial_y());
  set_attractor_state(trajectories[0].final_y());

  //-----------------------------------------------------
  // 2) Train a separate Dmp for each demonstration, and get the resulting model parameters
  std::set<std::string> selected;
  selected.insert("offsets");
  selected.insert("slopes");
  
  MatrixXd cur_task_parameters;
  VectorXd cur_model_parameters;// todo Remove redundant tmp variable
  vector<MatrixXd> all_model_parameters(n_demonstrations); 
  for (unsigned int i_demo=0; i_demo<n_demonstrations; i_demo++)
  {
    
    string save_directory_demo;
    if (!save_directory.empty())
      save_directory_demo = save_directory + "/demo" + to_string(i_demo);
    
    Dmp::train(trajectories[i_demo],save_directory_demo,overwrite);
    
    for (int i_dim=0; i_dim<dim_orig(); i_dim++)
    {

      // todo Should be argument of constructor
      function_approximator(i_dim)->setSelectedParameters(selected); 
  
      function_approximator(i_dim)->getParameterVectorSelected(cur_model_parameters);
      //cout << cur_model_parameters << endl;
      if (i_demo==0)
        all_model_parameters[i_dim].resize(n_demonstrations,cur_model_parameters.size());
      else
        assert(cur_model_parameters.size()==all_model_parameters[i_dim].cols());

      all_model_parameters[i_dim].row(i_demo) = cur_model_parameters;
      
    }
  }

   
  //-----------------------------------------------------
  // 3) Gather all task parameter values for all demonstrations
  
  // Gather task parameters in a matrix
  int n_task_parameters = task_parameters[0].cols();
  // This is the first time task_parameters_ is set, because this is the first time we know 
  // n_task_parameters.
  // We set it so that set_task_parameters can check if task_parameters_.cols()==n_task_parameters
  task_parameters_ = MatrixXd::Zero(1,n_task_parameters);
  VectorXd cur_task_parameters_t0;
  
  MatrixXd inputs(n_demonstrations,n_task_parameters);
  for (unsigned int i_demo=0; i_demo<n_demonstrations; i_demo++)
  {
    // These are the task parameters for the current demonstration at t=0
    cur_task_parameters_t0 = task_parameters[i_demo].row(0);
    
    // Task parameter may not change over time for 2-Step contextual DMP
    // Start comparison to i_time=0 at i_time=1
    for (int i_time=1; i_time<task_parameters[i_demo].rows(); i_time++)
    {
      if ( (cur_task_parameters_t0.array() != task_parameters[i_demo].row(i_time).array()).any())
      {
        cerr << __FILE__ << ":" << __LINE__ << ":";
        cerr << "WARNING. For DmpContextualTwoStep, task parameters may not vary over time during training. Using task parameters at t=0 only." << endl;
      }
    }

    // Take the first row, i.e. at time_i = 0. We checked above if they are constant over time.
    inputs.row(i_demo) = cur_task_parameters_t0;
  }

  //-----------------------------------------------------
  // 4) Train the policy parameter function for each dimension and each model parameter
  
  // Input to policy parameter functions: task_parameters
  // Target for each policy parameter function: all_model_parameters.col(param)
  
  for (int i_dim=0; i_dim<dim_orig(); i_dim++)
  {
    int n_pol_pars = all_model_parameters[i_dim].cols();
    for (int i_pol_par=1; i_pol_par<n_pol_pars; i_pol_par++)
    {
      policy_parameter_function_[i_dim].push_back(policy_parameter_function_[i_dim][0]->clone());
      //cout << *(policy_parameter_function_[i_dim][i_pol_par]) << endl;
    }

    for (int i_pol_par=0; i_pol_par<n_pol_pars; i_pol_par++)
    {
      MatrixXd targets = all_model_parameters[i_dim].col(i_pol_par);
      //cout << "_________________" << endl;
      //cout << inputs.transpose() << endl << endl;
      //cout << targets.transpose() << endl;
      
      string save_directory_cur;
      if (!save_directory.empty())
          save_directory_cur = save_directory + "/dim" + to_string(i_dim) + "_polpar" + to_string(i_pol_par);
      
      policy_parameter_function_[i_dim][i_pol_par]->train(inputs,targets,save_directory_cur,overwrite);
    
    }
  }
  
  
}

template<class Archive>
void DmpContextualTwoStep::serialize(Archive & ar, const unsigned int version)
{
  // serialize base class information
  ar & BOOST_SERIALIZATION_BASE_OBJECT_NVP(DmpContextual);
  
  ar & BOOST_SERIALIZATION_NVP(policy_parameter_function_);

}

}

File: dmpbbo/dmps/DmpContextualTwoStep.py
# This file is part of DmpBbo, a set of libraries and programs for the
# black-box optimization of dynamical movement primitives.
# Copyright (C) 2023 Freek Stulp
#
# DmpBbo is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# DmpBbo is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with DmpBbo.  If not, see <http://www.gnu.org/licenses/>.
#
""" Module for the DMP class. """
import copy

import matplotlib.pyplot as plt
import numpy as np
from matplotlib import cm

from dmpbbo.dmps.Dmp import Dmp
from dmpbbo.dynamicalsystems.DynamicalSystem import DynamicalSystem


class DmpContextualTwoStep(DynamicalSystem):
    """ Contextual Dynamical Movement Primitive

    Uses the two-step approach, i.e.
        1. determine DMP parameters from context parameters before integration
        2. integrate DMP with those parameters

    See the paper "Learning compact parameterized skills with a single regression" and
    ""Learning parameterized skills" for the nomenclature:
        https://ieeexplore.ieee.org/abstract/document/7030008/
        https://arxiv.org/abs/1206.6398
    """

    def __init__(self, tau, y_init, y_attr, dmp_function_apps, **kwargs):
        """Initialize a contextual DMP

        @param dmp: DMP to be parameterized by the context
        @param ppf_function_approximator: Function approximator for the policy parameter function
        """
        dim_y = len(y_init)
        dim_x = 3 * dim_y + 2
        dim_x += y_init.size  # new: damping coefficient system
        super().__init__(1, tau, y_init, dim_x)

        self.dmp_function_apps = dmp_function_apps
        self.dmp = Dmp(tau, y_init, y_attr, dmp_function_apps, **kwargs)

        # The policy parameter function can only be trained once trajectories have
        # been provided.
        self.ppf = None
        self._task_params_and_trajs_train = None

    def train(self, task_params_and_trajs, param_names, ppf_function_app, **kwargs):
        save_training_data = kwargs.get("save_training_data", False)

        self.dmp.set_selected_param_names(param_names)

        # Train the policy parameter function
        targets = []  # The dmp parameters
        inputs = []  # The task parameters
        taus = []
        y_attrs = []
        y_inits = []
        cur_dmp = None
        n_dmp_params = 0
        for task_param_and_traj in task_params_and_trajs:
            task_params = np.atleast_1d(task_param_and_traj[0])
            inputs.append(task_params)

            traj = task_param_and_traj[1]
            cur_dmp = copy.deepcopy(self.dmp)
            cur_dmp.train(traj)
            # h, axs = cur_dmp.plot(plot_demonstration=traj)
            dmp_params = cur_dmp.get_param_vector()
            n_dmp_params = len(dmp_params)
            targets.append(dmp_params)

            taus.append(cur_dmp.tau)
            y_attrs.append(cur_dmp.y_attr)
            y_inits.append(cur_dmp.y_init)

        self.dmp = cur_dmp
        self.dmp.tau = np.mean(taus)
        self.dmp.y_init = np.mean(y_inits, axis=0)
        self.dmp.y_attr = np.mean(y_attrs, axis=0)

        targets = np.array(targets)
        inputs = np.array(inputs)
        # Useful for plotting and sanity checks
        # self.task_params_train = inputs

        # ppf = policy parameter function
        if isinstance(ppf_function_app, list):
            if not len(ppf_function_app) == n_dmp_params:
                raise RuntimeError(
                    f"Length of 'ppf_function_app' list ({len(ppf_function_app)}) must be the as"
                    f" the vector of dmp parameters ({n_dmp_params}). "
                )
            self.ppf = ppf_function_app
        else:
            # Deep copies of separate function approximator, one for each DMP dim
            self.ppf = [copy.deepcopy(ppf_function_app) for _ in range(n_dmp_params)]

        for i_param in range(n_dmp_params):
            self.ppf[i_param].train(
                inputs, targets[:, i_param], save_training_data=save_training_data
            )

        self._task_params_and_trajs_train = task_params_and_trajs if save_training_data else None

    @classmethod
    def from_trajs(
        cls, task_params_and_trajs, dmp_function_apps, param_names, ppf_function_app, **kwargs
    ):
        """Initialize a DMP by training it from a trajectory.

        @param task_params_and_trajs: list of task_params / trajectory tuples
        @param dmp_function_apps: Function approximators for the DMP forcing term
        @param param_names: names of the function approximator parameters to predict.
        @param ppf_function_app: Function approximators for the policy-parameter function
        @param kwargs: All kwargs are passed to the Dmp constructor.
        """

        # Relevant variables from trajectory
        traj_index = 1
        first_traj = task_params_and_trajs[0][
            traj_index
        ]  # It is assumed tau, y_init and y_attr are the same for all
        tau = first_traj.ts[-1]
        y_init = first_traj.ys[0, :]
        y_attr = first_traj.ys[-1, :]

        dmp_contextual = cls(tau, y_init, y_attr, dmp_function_apps, **kwargs)

        dmp_contextual.train(task_params_and_trajs, param_names, ppf_function_app, **kwargs)

        return dmp_contextual

    @classmethod
    def from_dmp(cls, task_params_and_trajs, dmp, param_names, ppf_function_app, **kwargs):
        """Initialize a DMP by training it from a trajectory.

        @param task_params_and_trajs: list of task_params / trajectory tuples
        @param dmp: The DMP
        @param param_names: names of the function approximator parameters to predict.
        @param ppf_function_app: Function approximators for the policy-parameter function
        @param kwargs: All kwargs are passed to the Dmp constructor.
        """

        # Relevant variables from dmp
        fas = dmp._function_approximators  # noqa
        dmp_contextual = cls(dmp.tau, dmp.y_init, dmp.y_attr, fas)
        # Replace the dmp that was constructed in the above by the one that was passed.
        dmp_contextual.dmp = dmp

        dmp_contextual.train(task_params_and_trajs, param_names, ppf_function_app, **kwargs)

        return dmp_contextual

    def dim_dmp(self):
        return self.dmp.dim_dmp

    def set_task_params(self, task_params):
        dmp_params = self.dmp.get_param_vector()  # Get vector of right size; Values do not matter.
        for i_param, fa in enumerate(self.ppf):
            dmp_params[i_param] = fa.predict(task_params)
        self.dmp.set_param_vector(dmp_params)

    def analytical_solution(self, task_params, ts=None, suppress_forcing_term=False):
        self.set_task_params(task_params)
        return self.dmp.analytical_solution(ts, suppress_forcing_term)

    def integrate_start(self, task_params, y_init=None):
        """ Start integrating the DMP with a new initial state.

        @param task_params: The task parameters
        @param y_init: The initial state vector (y part)
        @return: x, xd - The first vector of state variables and their rates of change
        """

        # Predict dmp parameters from task parameters using the ppf
        n_params = self.dmp.get_param_vector_size()
        param_vector = np.zeros((n_params, 1))
        for i_param in range(n_params):
            param_vector[i_param] = self.ppf[i_param].predict(task_params)

        # Set the dmp parameters
        self.dmp.set_param_vector(param_vector)

        return self.dmp.integrate_start(y_init)

    def differential_equation(self, x):
        return self.dmp.differential_equation(x)

    def states_as_pos_vel_acc(self, x_in, xd_in):
        return self.dmp.states_as_pos_vel_acc(x_in, xd_in)

    def states_as_trajectory(self, ts, x_in, xd_in):
        return self.dmp.states_as_trajectory(ts, x_in, xd_in)

    def decouple_parameters(self):
        self.dmp.decouple_parameters()

    def plot_policy_parameter_function(self, **kwargs):
        axs = kwargs.get("axs", None)
        if axs is None:
            n_rows = int(np.ceil(np.sqrt(len(self.ppf))))
            _, axs = plt.subplots(n_rows, n_rows)
            axs = axs.flatten()
        hs = []
        for i_param, fa in enumerate(self.ppf):
            h, _ = fa.plot(ax=axs[i_param], **kwargs)
            hs.extend(h)
        return hs, axs

    def plot(self, task_params_and_trajs, **kwargs):

        axs = kwargs.get("axs", None)
        if axs is None:
            axs = self.dmp.get_dmp_axes()
            kwargs["axs"] = axs  # This will be passed to self.dmp later

        max_duration = 0.0
        dt = None
        all_task_params = []
        for task_param, traj_demo in task_params_and_trajs:
            max_duration = max(max_duration, traj_demo.duration)
            dt = traj_demo.dt_mean
            all_task_params.append(task_param)

            h_demo, _ = traj_demo.plot(axs=axs[1:4])
            plt.setp(h_demo, linestyle="-", linewidth=3, color=(0.7, 0.7, 0.7))

        ts = np.arange(0.0, 1.1 * max_duration, dt)

        all_task_params = np.array(all_task_params)
        tp_min = np.atleast_1d(np.min(all_task_params, axis=0))
        tp_max = np.atleast_1d(np.max(all_task_params, axis=0))

        hs = []
        cmap = cm.copper  # noqa
        prev_tau = self.dmp.tau
        for task_param, traj_demo in task_params_and_trajs:

            self.dmp.tau = traj_demo.duration
            self.dmp.y_init = traj_demo.y_init
            self.dmp.y_attr = traj_demo.y_final
            self.set_task_params(np.array([task_param]))

            h, _ = self.dmp.plot(ts, axs=axs, plot_no_forcing_term_also=True)

            tp = np.atleast_1d(task_param)
            scaled = (tp[0] - tp_min[0]) / (tp_max[0] - tp_min[0])
            color = cmap(np.clip(scaled, 0, 1))
            plt.setp(h, color=color)

            hs.extend(h)

            for ax in axs:
                ax.axvline(self.dmp.tau, color=color, linewidth=1)

        self.dmp.tau = prev_tau

        return hs, axs
